### PR TITLE
E2E Test Cases of eShopContainers Product Catalog 

### PR DESCRIPTION
**Pull request by augustocristian**
- Tests filtering functionalities, combining all the products Types and Brands
- Test Add-to-Basket functionality that checks that one-to several products can be added after the login, and this functionality is disabled when user is not logged in

### CODE REVIEW

**Comment by augustocristian** [comment]
Converted to draft to avoid notifications

File: src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+import static giis.eshopcontainers.e2e.functional.utils.Shopping.*;
+
+class CatalogTests extends BaseLoggedClass {
+    @Test
+    @DisplayName("AddProductsToBasket")
+    void addProductsToBasket() throws ElementNotFoundException {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java:17]: This test makes different steps (login, add to basket, logout). Readability would be improved if logical blocks are separated by a blank line

File: src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        log.debug("Checking that the product buttons are disabled before the user logs-in");
+        WebElement productCupButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[1]/form/input[1]"));
+        Assertions.assertEquals("esh-catalog-button is-disabled", productCupButton.getAttribute("class"), "Before Login the eShop product button was expected for being disabled and was enabled");
+        this.login();
+        WebElement loggedUser = driver.findElement(By.xpath("//*[@id=\"logoutForm\"]/section[1]/div"));
+        Assertions.assertEquals("alice", loggedUser.getText(), "The user logged is not the expected, the expected was alice and the obtained is " + loggedUser.getText());
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java:23]: Why alice is checked to be the logged user if the login didn't specified the user? Should this check be part of the login method who knows the user that is bein logged?

File: src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        log.debug("Adding the first element: NetCore Cup ");
+        productCupButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[1]/form/input[1]"));
+        Assertions.assertEquals("esh-catalog-button ", productCupButton.getAttribute("class"), "The eShop product button was expected for being enabled and was disabled");
+        Click.element(driver, waiter, productCupButton);
+        Assertions.assertEquals(1, getNumShoppingItems(driver, waiter), "The number of elements in the basket doesn't match");
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java:28]: This block is for adding an item, there are two more similar blocks after this. Could be generalized this operation using a private function to add an item to basket?

File: src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        log.debug("Adding Two elements more: Hoodie and CupPin");
+        WebElement productHoodieBlackButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[3]/form/input[1]"));
+        Click.element(driver, waiter, productHoodieBlackButton);
+        Assertions.assertEquals(2, getNumShoppingItems(driver, waiter));
+        WebElement productCupPinButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[6]/form/input[1]"));
+        Click.element(driver, waiter, productCupPinButton);
+        Assertions.assertEquals(3, getNumShoppingItems(driver, waiter));
+        this.logout();
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java:36]: I have a concern about the login/logout pattern. During the first test case review, I got the idea that the base class (BaseLoggedClass) was created for every test that requires a logged user. However, this case seems to be intended to browse to the basket page without a login, then loging, and then add to basket and logout. But what would happen if other test case does a similar scenario and fails in the middle? Could let the user logged in and then enter in this test with an already logged user?

Should the logout be inside a teardown method (in all tests) to ensure that this situation does not happens?

File: src/test/java/giis/eshopcontainers/e2e/functional/utils/Shopping.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class with various methods facilitating the catalog-shopping actions
+ */
+public class Shopping {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/utils/Shopping.java:15]: Package utils contains general purpose support classes, like navigation, clicking, etc. However this class is domain specific, more precisely, test specific (about shopping cart). Methods here should be part of the test that is using them (at least, in view of the tests that we have now)

File: src/test/java/giis/eshopcontainers/e2e/functional/utils/Shopping.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        log.debug("Selecting the Brand: " + new String[]{"All Brands", "Net Core", "Others"}[option - 1]);
+        Click.element(driver, waiter, elementOfInterest);
+    }
+
+    /**
+     * This method applies several type filters to the eShopOnContainers catalog
+     * @param option Type of filter: 1) All Types, 2) Mug, 3) TShirt and 4)Pin
+     */
+    public static void selectTypeFilter(WebDriver driver, Waiter waiter, Integer option) throws ElementNotFoundException {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/utils/Shopping.java:49]: This method is a nearly exact duplicate of  `selectBrandFilter`, both methods should be generalized into only one.

**Review by javiertuya — CHANGES_REQUESTED** [review]

File: src/test/java/giis/eshopcontainers/e2e/functional/utils/Shopping.java (hunk 6 above)
Comment (augustocristian) [reply src/test/java/giis/eshopcontainers/e2e/functional/utils/Shopping.java:49]: I have reformatted the code, and the new methods now share a common one named "applyFilter," which encapsulates the core functionality. I believe having two distinct methods enhances the readability and understanding of the test case.

**Comment by augustocristian** [comment]
Some discussion:
Regarding the BaseLoggedTest, the setup might be preferable as follows:
``` @BeforeEach
void setup(TestInfo testInfo) {
    log.info("Starting Individual Set-up for the test: {}.", testInfo.getDisplayName());

    initializeDriverAndWaiter();
    retrieveTestJobName();
    retrieveUserCredentials();
    navigateToSutUrl();

    log.info("Individual Set-up finished, starting test: {}.", testInfo.getDisplayName());
}

private void initializeDriverAndWaiter() {
    log.debug("Initializing WebDriver and Waiter instances.");
    driver = seleManager.getDriver();
    waiter = new Waiter(driver);
}

private void retrieveTestJobName() {
    tjobName = System.getProperty("tjob_name");
}

private void retrieveUserCredentials() {
    userName = properties.getProperty("USER_ESHOP");
    password = properties.getProperty("USER_ESHOP_PASSWORD");
}

private void navigateToSutUrl() {
    log.debug("Navigating to {}", sutUrl);
    driver.get(sutUrl);
}  ```
That may ease maintainability (but now in my opinion looks too complex) or the current one (I've added some comments to explain what actions are performed and ease its understanding,  especially when revisiting the code after some time

File: src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java (hunk 4 above)
Comment (augustocristian) [reply src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java:36]: Javier, currently the logout its also in the teardown method (if the test fails for any reason, it goes to the tear-down method that performs:
```
    @AfterEach
    void tearDown(TestInfo testInfo) throws ElementNotFoundException {
        log.info("Disposing user and releasing/closing browser for the test {}", testInfo.getDisplayName());
        if (isLogged) {
            log.debug("Logging out user: {}", this.userName);
            this.logout();
        }
    }
```

**Comment by javiertuya** [comment]
> Some discussion:
> Regarding the BaseLoggedTest, the setup might be preferable as follows:

Let postpone the discussion to a separate issue

**Review by javiertuya — APPROVED** [review]
Again including changes that are not related to the review comments !!!!

**Pull request merged**